crux-sql/src/crux/calcite.clj

```diff
@@ -0,0 +1,487 @@
+
+      (when-let [custom-fn (and (= SqlKind/OTHER_FUNCTION (.getKind n)) (crux-custom-fns (str (.-op ^RexCall n))))]
+        (custom-fn (.getValue2 ^RexLiteral (first (.-operands ^RexCall n)))))
+
+      (when-let [op (get arithmetic-fns (.getKind n))]
+        (ArbitraryFn. op (map #(->ast % schema) (.getOperands ^RexCall n))))
+
+      (and (#{SqlKind/AND SqlKind/OR SqlKind/NOT} (.getKind n))
+           (SQLCondition. (.getKind n) (mapv #(->ast % schema) (.-operands ^RexCall n))))
+
+      (and (instance? RexCall n) (linq-lambda n schema))
+
+      n))
+
+(defn- ->vars+clauses+exprs
+  "Swaps out Calcs/Predicates for symbols and clauses."
+  [schema nodes]
+  (let [args (atom {})
+        exprs (atom {})
+        clauses (atom [])]
+    [(postwalk (fn [x]
+                 (condp instance? x
+
+                   Expression
+                   (let [s (gensym)]
+                     (swap! exprs assoc s x)
+                     s)
+
+                   ArbitraryFn
+                   (let [s (gensym)]
+                     (swap! clauses conj [(apply list (.op ^ArbitraryFn x) (.operands ^ArbitraryFn x)) s])
+                     s)
+
+                   RexDynamicParam
+                   (let [sym (gensym)]
+                     (swap! args assoc (.getName ^RexDynamicParam x) sym)
+                     sym)
+
+                   RexLiteral
+                   (let [s (gensym)]
+                     (swap! exprs assoc s (->literal-expression x))
+                     s)
+
+                   RexInputRef
+                   (get-in schema [:crux.sql.table/query :find (.getIndex ^RexInputRef x)])
+
+                   x))
+               nodes)
+     @clauses @exprs @args]))
+
+(defn- ground-vars [or-statement]
+  (let [vars (distinct (mapcat #(filter symbol? (rest (first %))) or-statement))]
+    (vec
+     (for [clause or-statement]
+       (apply list 'and clause  (map #(vector (list 'identity %)) vars))))))
+
+(defn ->where [x]
+  (condp instance? x
+
+    SQLPredicate
+    [(apply list (.op ^SQLPredicate x) (map #(if (symbol? %) % (->where %))
+                                            (.operands ^SQLPredicate x)))]
+
+    SQLCondition
+    (condp = (.c ^SQLCondition x)
+      SqlKind/AND
+      (mapv ->where (.clauses ^SQLCondition x))
+      SqlKind/OR
+      (apply list 'or (ground-vars (map ->where (.clauses ^SQLCondition x))))
+      SqlKind/NOT
+      (apply list 'not (map ->where (.clauses ^SQLCondition x))))
+
+    Symbol
+    [(list '= x true)]
+
+    ;; I.e. keywords, uuids
+    x))
+
+(defn- vectorize-value [v]
+  (if (or (not (vector? v)) (= 1 (count v))) (vector v) v))
+
+(defn enrich-filter [schema ^RexNode filter]
+  (log/debug "Enriching with filter" filter)
+  (let [[filters clauses exprs args] (->vars+clauses+exprs schema (->ast filter schema))]
+    (-> schema
+        (update-in [:crux.sql.table/query :where] (comp vec concat) (vectorize-value (->where filters)))
```

Comment:
nit: can do this with `(comp vec concat)` replaced with `(fnil into [])` (or just `into` if you know it's present) - saves going to a lazy-seq and back again

Reply:
went with `fnil`